cmd/util/common/checkpoint.go

```diff
@@ -0,0 +1,246 @@
+	logger zerolog.Logger,
+	state protocol.State,
+	headers storage.Headers,
+	seals storage.Seals,
+	checkpointDir string,
+	blocksToSkip uint,
+	endHeight uint64,
+) (protocol.Snapshot, uint64, flow.StateCommitment, error) {
+	startHeight := uint64(0)
+	// preventing startHeight from being negative
+	length := uint64(blocksToSkip+1) * 10000
+	if endHeight > length {
+		startHeight = endHeight - length
+	}
+
+	checkpointFilePath, err := findLatestCheckpointFilePath(checkpointDir)
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, fmt.Errorf("could not find latest checkpoint file in directory %v: %w", checkpointDir, err)
+	}
+
+	log.Info().
+		Uint64("start_height", startHeight).
+		Uint64("end_height", endHeight).
+		Uint("blocksToSkip", blocksToSkip).
+		Msgf("generating protocol snapshot for checkpoint file %v", checkpointFilePath)
+	// find the height of the finalized block that produces the state commitment contained in the checkpoint file
+	sealedHeight, commit, err := FindHeightByCheckpoints(logger, headers, seals, checkpointFilePath, blocksToSkip, startHeight, endHeight)
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, fmt.Errorf("could not find sealed height in range [%v:%v] (blocksToSkip: %v) by checkpoints: %w",
+			startHeight, endHeight, blocksToSkip,
+			err)
+	}
+
+	// find which finalized height seals the block with the given sealed height
+	finalizedHeight, err := findFinalizedHeightBySealedHeight(state, sealedHeight)
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, fmt.Errorf("could not find finalized height for sealed height %v: %w", sealedHeight, err)
+	}
+
+	snapshot := state.AtHeight(finalizedHeight)
+	validSnapshot, err := snapshots.GetDynamicBootstrapSnapshot(state, snapshot)
+	if err != nil {
+		return nil, 0, flow.DummyStateCommitment, fmt.Errorf("could not get dynamic bootstrap snapshot: %w", err)
+	}
+
+	return validSnapshot, sealedHeight, commit, nil
+}
+
+// hashesToCommits converts a list of ledger.RootHash to a list of flow.StateCommitment
+func hashesToCommits(hashes []ledger.RootHash) []flow.StateCommitment {
+	commits := make([]flow.StateCommitment, len(hashes))
+	for i, h := range hashes {
+		commits[i] = flow.StateCommitment(h)
+	}
+	return commits
+}
+
+// findFinalizedHeightBySealedHeight finds the finalized height that seals the block with the given sealed height
+func findFinalizedHeightBySealedHeight(state protocol.State, sealedHeight uint64) (uint64, error) {
+	sealed, err := state.AtHeight(sealedHeight).Head()
+	if err != nil {
+		return 0, err
+	}
+	sealedID := sealed.ID()
+
+	lastFinalized, err := state.Final().Head()
+	if err != nil {
+		return 0, fmt.Errorf("could not get last finalized block: %w", err)
+	}
+	lastFinalizedHeight := lastFinalized.Height
+
+	// the finalized height that seals the given sealed height must be above the sealed height
+	// so if we iterate through each height, we should eventually find the finalized height
```

Comment:
Good catch and good idea!

Reply:
I think it's simple here that we don't need to implement this function, but I think it could be a useful utility function to know which block finalizes the seal for a given block. We don't have an index for this type of lookup, and the search is currently missing.  

The scenario could be: I submitted a transaction A, which is included in block 10, I would like to know which block seals my transaction (seals block 10).

Is there any other scenario that needed this function?